### DISCUSSION TITLE
How to evaluate multiple frames with the C++ interface?

### DISCUSSION THREAD

**Discussion by cihe13375** [deepmodeling/deepmd-kit#2214] [Q&A] [answered]
In C++ API doc (e.g. [link](https://docs.deepmodeling.com/projects/deepmd/en/v2.1.5/API_CC/classdeepmd_1_1DeepPot.html#_CPPv4N6deepmd7DeepPot7computeER10ENERGYTYPERNSt6vectorI9VALUETYPEEERNSt6vectorI9VALUETYPEEERKNSt6vectorI9VALUETYPEEERKNSt6vectorIiEERKNSt6vectorI9VALUETYPEEERKNSt6vectorI9VALUETYPEEERKNSt6vectorI9VALUETYPEEE)) it's stated that the `coord` and `box` parameters support multiple frame input (i.e. size of nframes x natoms x 3 and nframes x 9, respectively). So I tried to do something like this:

```
// mostly copied from the example
#include "deepmd/DeepPot.h"
#include <vector>
#include <iostream>

int main(){
  deepmd::DeepPot dp("forcefield.pb");
  // 2 identical frames
  std::vector<float> coord = {1., 0., 0., 0., 0., 1.5, 1. ,0. ,3., 1., 0., 0., 0., 0., 1.5, 1. ,0. ,3.};
  std::vector<float> cell= {10., 0., 0., 0., 10., 0., 0., 0., 10., 10., 0., 0., 0., 10., 0., 0., 0., 10.};
  std::vector<int> atype = {1, 0, 1};
  double e;
  std::vector<float> f, v;
  dp.compute (e, f, v, coord, atype, cell);
  std::cout << f[0] << std::endl;
  return 0;
} 
```
and then compile with:
`g++ filename.cpp -L/path/to/dpmd/lib -I/path/to/dpmd/include -ldeepmd_cc_low -lstdc++ -ltensorflow_cc`
It compiles, but crash at runtime with `terminate called after throwing an instance of 'std::length_error'
  what():  vector::_M_range_insert` (if there're more frames it directly segfaults).
If I also replicate atype (i.e. `std::vector<int> atype = {1, 0, 1, 1, 0, 1};`) it runs, but the force is NaN.
I also tried nested vectors (e.g. `std::vector<std::vector<float>> coord(2, {1., 0., 0., 0., 0., 1.5, 1. ,0. ,3.});` etc.) but it doesn't compile due to `no matching function for call to ...` (it seems that `VALUETYPE` must be `float` and cannot be another `std::vector`).

So I'm wondering how to use the C++ interface correctly with multiple frame input, or this is not implemented at all?
I'm using version 2.1.5-cuda11.6_gpu installed by the offline package. The python interface seems to be fine (just pass in a e.g. nframes x natoms x 3 numpy array and it works), and the C++ interface is fine with single-frame input.

**Comment by njzjz** [comment, marked as answer]
Yes, only one frame is supported. The documentation is wrong.

**Reply by cihe13375** [reply]
documentation is still unfixed :(

**Reply by njzjz** [reply]
Contribution is welcome.

**Comment by njzjz** [comment]
It's now supported by #2305.
